javaslang/src/main/java/javaslang/collection/AbstractsQueue.java

```diff
@@ -0,0 +1,100 @@
+package javaslang.collection;
+
+import javaslang.*;
+import javaslang.control.Option;
+
+import java.util.*;
+
+abstract class AbstractsQueue<T, Q extends AbstractsQueue<T, Q>> implements Kind1<Q, T>, LinearSeq<T> {
+    private static final long serialVersionUID = 1L;
+
+    /**
+     * Removes an element from this Queue.
+     *
+     * @return a tuple containing the first element and the remaining elements of this Queue
+     * @throws NoSuchElementException if this Queue is empty
+     */
+    abstract Tuple2<T, Q> dequeue();
+
+    /**
+     * Removes an element from this Queue.
+     *
+     * @return {@code None} if this Queue is empty, otherwise {@code Some} {@code Tuple} containing the first element and the remaining elements of this Queue
+     */
+    public Option<Tuple2<T, Q>> dequeueOption() {
+        return isEmpty() ? Option.none() : Option.some(dequeue());
+    }
+
+    /**
+     * Enqueues a new element.
+     *
+     * @param element The new element
+     * @return a new {@code Queue} instance, containing the new element
+     */
+    public abstract Q enqueue(T element);
+
+    /**
+     * Enqueues the given elements. A queue has FIFO order, i.e. the first of the given elements is
+     * the first which will be retrieved.
+     *
+     * @param elements Elements, may be empty
+     * @return a new {@code Queue} instance, containing the new elements
+     * @throws NullPointerException if elements is null
+     */
+    @SuppressWarnings("unchecked")
+    public Q enqueue(T... elements) {
+        Objects.requireNonNull(elements, "elements is null");
+        return enqueueAll(List.of(elements));
+    }
+
+    /**
+     * Enqueues the given elements. A queue has FIFO order, i.e. the first of the given elements is
+     * the first which will be retrieved.
+     *
+     * @param elements An Iterable of elements, may be empty
+     * @return a new {@code Queue} instance, containing the new elements
+     * @throws NullPointerException if elements is null
+     */
+    @SuppressWarnings("unchecked")
+    public Q enqueueAll(Iterable<? extends T> elements) {
+        Objects.requireNonNull(elements, "elements is null");
+
+        return List.ofAll(elements).foldLeft((Q) this, AbstractsQueue<T, Q>::enqueue); // TODO crashes the compiler for Q::enqueue
```

Comment:
Does it crash `javac`, ECJ or the IntelliJ error marker?

If it crashes `javac` we have to find a workaround.

If it crashes the Eclise Compiler or IntelliJ error marker, we add bug-reports on the bug trackers.